### PR TITLE
#2 Convert Bulk API library users upsert to use a query as the id is gone

### PR DESCRIPTION
**Pull request by jon-betts**
Now we don't have an id for the user, just saying "upsert" doesn't really make any sense. We need to state _what_ we are upserting by. This now follows the pattern laid out by the group which is to separate fields sufficient to uniquely identify the object into `query` (which are automatically folded into the attributes for insertion).

The difference is most easily seen in the schema examples:

 * https://github.com/hypothesis/h/pull/5982/files#diff-7c170d7c5bcaaee382c64aab2d33c9b0

In short:

 * "username" and "authority" are now part of the "query" rather than attributes for transit
 * As we have "merge_queries" enabled, this will get copied into the attributes for convenience on the other side
 * This lets us specify how we want to find the user to update
 * This correspondingly lets us check on the other side that the query is what we expect (this is the only mechanism we support)

### CODE REVIEW

File: h/h_api/resources/schema/core.json [hunk 1]
```diff
@@ -50,24 +50,28 @@
             ]
         },
 
+        "userName": {
+            "type": "string",
+            "minLength": 3,
+            "maxLength": 30,
+            "pattern": "^[A-Za-z0-9._]+$"
+        },
```
Comment (jon-betts) [inline h/h_api/resources/schema/core.json]: As we re-use the `userName` in the attrs and query specification, split it out

File: tests/h/h_api/bulk_api/entry_point_test.py [hunk 2]
```diff
@@ -74,14 +75,18 @@ def test_we_reject_bad_arguments(self, method, kwargs, executor):
...
-        self._assert_nd_json_matches_commands(handle.getvalue(), commands)
+        assert self._decode_ndjson(handle.getvalue()) == expected
 
     def test_to_string(self, commands):
+        expected = deepcopy([command.raw for command in commands])
+
         nd_json = BulkAPI.to_string(commands)
 
-        self._assert_nd_json_matches_commands(nd_json, commands)
+        assert self._decode_ndjson(nd_json) == expected
```
Comment (jon-betts) [inline tests/h/h_api/bulk_api/entry_point_test.py]: Due to query merging, we need to capture the expected output before we process the queries (as the post merge isn't what we send)

File: tests/h/h_api/conftest.py [hunk 3]
```diff
@@ -59,12 +59,18 @@ def create_group_membership_body():
 
 @pytest.fixture
 def user_attributes(upsert_user_body):
-    return upsert_user_body["data"]["attributes"]
+    return dict(
+        upsert_user_body["data"]["attributes"],
+        **upsert_user_body["data"]["meta"]["query"]
```
Comment (jon-betts) [inline tests/h/h_api/conftest.py]: Merge the query into the attrs to get the items we need to call the create method with
